### ISSUE TITLE
Distributed training problem

### ISSUE THREAD

**Issue by Mrhs121** [tensorflow/tensorflow#40586]
I am studying the principle of distributed training of TensorFlow, but I cannot find the source code related to gradient communication. I want to know where can I find the source code of uploaded gradient to PS.

**Comment by nikitamaia** [comment]
Hi @Mrhs121, if I understand correctly it sounds like you are wondering how the gradients are updated and synchronized. Is there a specific distribution strategy you are curious about?

**Comment by Mrhs121** [comment]
yes

**Comment by nikitamaia** [comment]
There are different ways to manage the gradients in tf.distribute, depending on what strategy you are using and the arguments you configure. 

I would recommend first taking a look at the [overview page in the documentation](https://www.tensorflow.org/guide/distributed_training) to get a sense for the different options available. This [deep dive presentation](https://www.youtube.com/watch?v=jKV53r9-H14) might also be of use.

You can easily see the source code in the documentation as well. For example, if you were interested in learning about the tf.distribute implementation of [NcclAllReduce](https://www.tensorflow.org/api_docs/python/tf/distribute/NcclAllReduce), you can select the "view source on github" option to see the [source code](https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/distribute/cross_device_ops.py#L743-L765).

I will close this issue now since it is not a bug, and Stack Overflow is a better place to get support help with a larger community. But feel free to file a new issue if you run into bugs/issues in the future implementing any of these strategies.
